packages/react/menubar/src/Menubar.tsx

```diff
@@ -242,9 +325,50 @@ const MenubarContent = React.forwardRef<MenubarContentElement, MenubarContentPro
           // Always prevent auto focus because we either focus manually or want user agent focus
           event.preventDefault();
         })}
+        onFocusOutside={composeEventHandlers(props.onFocusOutside, (event) => {
+          const target = event.target as HTMLElement;
+          const isMenubarTrigger = getItems().some((item) => item.ref.current?.contains(target));
+          if (isMenubarTrigger) event.preventDefault();
+        })}
         onInteractOutside={composeEventHandlers(props.onInteractOutside, () => {
           hasInteractedOutsideRef.current = true;
         })}
+        onEntryFocus={(event) => {
+          if (!menuContext.wasKeyboardTriggerOpenRef.current) event.preventDefault();
+        }}
+        onKeyDown={composeEventHandlers(
+          props.onKeyDown,
+          (event) => {
+            if (['ArrowRight', 'ArrowLeft'].includes(event.key)) {
+              const target = event.target as HTMLElement;
+              const targetIsSubTrigger = target.hasAttribute('aria-haspopup');
+              const isKeyDownInsideRoot =
+                target.closest('[data-radix-menu-content]') === event.currentTarget;
```

Comment:
These two selections are controversial as we're relying on another components implementation details. I do think there is value here though as it allows us to colocate the entirely of this logic here, would providing attrs specific to this component alleviate some of the concerns?

Reply:
> These two selections are controversial as we're relying on another components implementation details. I do think there is value here though as it allows us to colocate the entirely of this logic here,

I agree with the premise.

> would providing attrs specific to this component alleviate some of the concerns?

Given what we've recently gone through with [this](https://github.com/radix-ui/primitives/pull/1723) I'd say yes.